### PR TITLE
LA-61: Copy multiple files to Shared Space

### PR DESCRIPTION
**Pull request by manhduc-nguyen**
Please review this
Screen Shot 2021-01-14 at 15 29 32 Screen Shot 2021-01-14 at 15 29 39

### CODE REVIEW

File: domain/lib/src/usecases/shared_space/copy_multiple_files_to_shared_space_interactor.dart [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+      {@required List<Document> documents,
+      @required SharedSpaceId destinationSharedSpaceId,
+      WorkGroupNodeId destinationParentNodeId}) async {
+    final listResult = <Either<Failure, Success>>[];
+    await Future.wait(documents.map((element) =>
+            _copyDocumentsToSharedSpaceInteractor.execute(
+                CopyRequest(element.documentId.uuid, SpaceType.personalSpace),
+                destinationSharedSpaceId,
+                destinationParentNodeId: destinationParentNodeId)))
+        .then((results) => listResult.addAll(results));
```
Comment (hoangdat) [inline domain/lib/src/usecases/shared_space/copy_multiple_files_to_shared_space_interactor.dart:59]: why not 
```
        final listResult = await Future.wait(documents.map((element) =>
            _copyDocumentsToSharedSpaceInteractor.execute(
                CopyRequest(element.documentId.uuid, SpaceType.personalSpace),
                destinationSharedSpaceId,
                destinationParentNodeId: destinationParentNodeId)));
```

File: domain/test/usecases/shared_space/copy_multiple_files_to_shared_space_interactor_test.dart [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+      sharedSpaceDocumentRepository = MockSharedSpaceDocumentRepository();
+      copyDocumentsToSharedSpaceInteractor = CopyDocumentsToSharedSpaceInteractor(sharedSpaceDocumentRepository);
+      copyMultipleFilesToSharedSpaceInteractor = CopyMultipleFilesToSharedSpaceInteractor(copyDocumentsToSharedSpaceInteractor);
+    });
+
+    test('copy to shared space interactor should return success with valid data', () async {
+      when(sharedSpaceDocumentRepository.copyToSharedSpace(
+          CopyRequest(document1.documentId.uuid, SpaceType.personalSpace),
+          sharedSpace1.sharedSpaceId))
+          .thenAnswer((_) async => [workGroupDocument1]);
```
Comment (hoangdat) [inline domain/test/usecases/shared_space/copy_multiple_files_to_shared_space_interactor_test.dart:56]: so hard to understand with that indent

File: domain/test/usecases/shared_space/copy_multiple_files_to_shared_space_interactor_test.dart [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+          CopyRequest(document2.documentId.uuid, SpaceType.personalSpace),
+          sharedSpace1.sharedSpaceId))
+          .thenAnswer((_) async => [workGroupDocument1]);
+
+
+      final result = await copyMultipleFilesToSharedSpaceInteractor.execute(
+          documents: [document1, document2],
+          destinationSharedSpaceId: sharedSpace1.sharedSpaceId);
+      final state = result.getOrElse(() => null);
+      expect(state, isA<CopyMultipleFilesToSharedSpaceAllSuccessViewState>());
```
Comment (hoangdat) [inline domain/test/usecases/shared_space/copy_multiple_files_to_shared_space_interactor_test.dart:67]: I also to verify the items of the response when copy success.

File: domain/test/usecases/shared_space/copy_multiple_files_to_shared_space_interactor_test.dart [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+          CopyRequest(document2.documentId.uuid, SpaceType.personalSpace),
+          sharedSpace1.sharedSpaceId))
+          .thenThrow(Exception());
+
+
+      final result = await copyMultipleFilesToSharedSpaceInteractor.execute(
+          documents: [document1, document2],
+          destinationSharedSpaceId: sharedSpace1.sharedSpaceId);
+      final state = result.getOrElse(() => null);
+      expect(state, isA<CopyMultipleFilesToSharedSpaceHasSomeFilesFailedViewState>());
```
Comment (hoangdat) [inline domain/test/usecases/shared_space/copy_multiple_files_to_shared_space_interactor_test.dart:85]: idem

File: lib/presentation/view/header/more_action_bottom_sheet_header_builder.dart [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+import 'package:filesize/filesize.dart';
+import 'package:flutter/material.dart';
+import 'package:linshare_flutter_app/presentation/localizations/app_localizations.dart';
+import 'package:linshare_flutter_app/presentation/model/file/presentation_file.dart';
+import 'package:linshare_flutter_app/presentation/util/extensions/color_extension.dart';
+
+class MoreActionBottomSheetHeaderBuilder {
+  final Key _key;
+  final List<PresentationFile> _files;
```
Comment (hoangdat) [inline lib/presentation/view/header/more_action_bottom_sheet_header_builder.dart:41]: I don't think it is good for this case, how about to transfer which files is operated in the future?

File: lib/presentation/widget/myspace/my_space_viewmodel.dart [hunk 6]
```diff
@@ -114,8 +115,8 @@ class MySpaceViewModel extends BaseViewModel {
     store.dispatch(_exportFileAction(document, cancelToken));
   }
 
-  void shareDocuments(List<Document> documents) {
-    if (documents.length == 1) {
+  void shareDocuments(List<Document> documents, {bool isFromMultipleSelection = false}) {
```
Comment (hoangdat) [inline lib/presentation/widget/myspace/my_space_viewmodel.dart:118]: magic parameter. No. It is hard to understand, Why we need this?

File: lib/presentation/widget/myspace/my_space_viewmodel.dart [hunk 7]
```diff
@@ -126,8 +127,11 @@ class MySpaceViewModel extends BaseViewModel {
             documents: documents));
   }
 
-  void copyToAWorkgroup(BuildContext context, Document document) {
+  void copyToAWorkgroup(BuildContext context, List<Document> documents, {bool isFromMultipleSelection = false}) {
```
Comment (hoangdat) [inline lib/presentation/widget/myspace/my_space_viewmodel.dart:130]: idem

File: lib/presentation/widget/myspace/my_space_viewmodel.dart [hunk 8]
```diff
@@ -293,6 +297,16 @@ class MySpaceViewModel extends BaseViewModel {
     store.dispatch(MySpaceClearSelectedDocumentsAction());
   }
 
+  void openMoreActionBottomMenu(BuildContext context, List<Document> documents, List<Widget> actionTiles) {
+    ContextMenuBuilder(context)
+        .addHeader(MoreActionBottomSheetHeaderBuilder(
+        context,
+        Key('more_action_menu_header'),
```
Comment (hoangdat) [inline lib/presentation/widget/myspace/my_space_viewmodel.dart:304]: indent

**Comment by manhduc-nguyen** [comment]
@hoangdat please check again.

**Review by hoangdat — APPROVED** [review]

**Pull request merged**